### PR TITLE
build(): First version of using SauceLabs and BrowserStack on travis.

### PR DESCRIPTION
**Pull request by hansl**

### CODE REVIEW

**Comment by hansl** [comment]
@jelbourn Just rerunning the build this morning was enough to make it pass. Yesterday's last test was a flake. Not 100% sure how I feel about that. Build 37 is the previous one that passed this morning.

**Comment by jelbourn** [comment]
@hansl We can go with it for now and see if it continues to be flaky. Can you squash your commits? 

cc @mlaval: Marc, if you have a few minutes to take a look that would be much appreciated.

File: test/browser-providers.conf.js [hunk 1]
```diff
@@ -0,0 +1,331 @@
...
+function decode(str) {
+  return (str || '').split('').reverse().join('');
+}
+
+if (process.env.TRAVIS) {
+  process.env.SAUCE_ACCESS_KEY = decode(process.env.SAUCE_ACCESS_KEY);
+  process.env.BROWSER_STACK_ACCESS_KEY = decode(process.env.BROWSER_STACK_ACCESS_KEY);
+}
+
+function buildConfiguration(type, target, required) {
```
Comment (jelbourn) [inline test/browser-providers.conf.js]: Could you add a brief function description comment?

File: test/karma.conf.js [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+    browsers: ['Chrome'],
+
+    singleRun: false
+  });
+
+  if (process.env.TRAVIS) {
+    var buildId = `TRAVIS #${process.env.TRAVIS_BUILD_NUMBER} (${process.env.TRAVIS_BUILD_ID})`;
+
+    var modeSplit = process.env.MODE.split('_');
+    var platform = modeSplit[0];
```
Comment (jelbourn) [inline test/karma.conf.js]: Can you add a comment for what `process.env.MODE` should look like?

**Comment by jelbourn** [comment]
Also FYI @IgorMinar @juliemr

File: test/karma.conf.js [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+    ],
+    files: [
+      {pattern: 'node_modules/es6-shim/es6-shim.js', included: true, watched: true},
+      {pattern: 'node_modules/angular2/bundles/angular2-polyfills.js', included: true, watched: true},
+      {pattern: 'node_modules/systemjs/dist/system-polyfills.js', included: true, watched: true},
+      {pattern: 'node_modules/systemjs/dist/system.src.js', included: true, watched: true},
+      {pattern: 'node_modules/rxjs/bundles/Rx.js', included: true, watched: true},
+      {pattern: 'node_modules/angular2/bundles/angular2.js', included: true, watched: true},
+      {pattern: 'node_modules/angular2/bundles/testing.dev.js', included: true, watched: true},
+      {pattern: 'node_modules/angular2/bundles/testing.dev.js', included: true, watched: true},
```
Comment (juliemr) [inline test/karma.conf.js]: This is duplicated?

**Comment by jelbourn** [comment]
LGTM, but I'd wait a couple hours in case someone else wants to comment on something.

File: test/karma.conf.js [hunk 4]
```diff
@@ -0,0 +1,32 @@
+// This file only hook up on require calls to transpile the TypeScript.
+// If you're looking at this file to see Karma configuration, you should look at
+// karma.config.ts instead.
```
Comment (jelbourn) [inline test/karma.conf.js]: After seeing this, my feelings are pretty mixed. It adds this extra layer of indirection and isn't what people expect when looking at a `karma.conf.js`. And the code in the config file is simple enough that you don't get a huge benefit from TS.

@robertmesserle What do you think?

Comment (hansl) [inline test/karma.conf.js]: In my defense, the browser-providers files is getting smaller and easier to read. The configuration itself doesn't fully benefit but might in the future if we add stuff.

I agree though that this is a small section of the repo that won't get touched regularly.

File: test/browser-providers.ts [hunk 5]
```diff
@@ -0,0 +1,349 @@
...
+}
+
+
+/** Setup the access keys */
+if (process.env.TRAVIS) {
+  process.env.SAUCE_ACCESS_KEY = decode(process.env.SAUCE_ACCESS_KEY);
+  process.env.BROWSER_STACK_ACCESS_KEY = decode(process.env.BROWSER_STACK_ACCESS_KEY);
+}
+
+/** Build a configuration. */
```
Comment (jelbourn) [inline test/browser-providers.ts]: This comment doesn't actually say anything (e.g., what a configuration consists of and what consumes it)

Comment (hansl) [inline test/browser-providers.ts]: Done.

**Comment by hansl** [comment]
Found the problem with the last few updates; the browserStack didn't properly start a tunnel and one was needed. I set the `startTunnel` var to `true` in the karma config for browserStack and it works.

Now, angular2 has a script to manually start a tunnel. I'm not 100% what the advantage of having this kind of control for us would be, but I'll consider it in a future update.

For now, I consider this PR done and working and will rebase and merge tomorrow morning, unless @jelbourn wants to put this on hold to review.

**Comment by IgorMinar** [comment]
We start tunnel via a script because we want to reuse one tunnel across
many invocations of karma within the same build job.

Additionally since the tunnel takes a while to start we start initiating it
at the beginning of the job, then build stuff and by the time karma needs
to run the tunnel is ready.

I suggest you stick to the same setup as A2

File: scripts/browserstack/start_tunnel.js [hunk 6]
```diff
@@ -0,0 +1,50 @@
+'use strict';
```
Comment (jelbourn) [inline scripts/browserstack/start_tunnel.js]: Could you add comment up top here explaining what the purpose of this script is?

File: src/components/button/button.ts [hunk 7]
```diff
@@ -29,7 +29,7 @@ export class MdButton {
     // button continues to look :active after clicking.
     // @see http://marcysutton.com/button-focus-hell/
     this.isMouseDown = true;
-    setTimeout(() => {this.isMouseDown = false}, 100);
+    setTimeout(() => { this.isMouseDown = false; }, 100);
```
Comment (jelbourn) [inline src/components/button/button.ts]: Need to rebase?

Comment (hansl) [inline src/components/button/button.ts]: I did rebase, not sure what's going on. Let me play god with history a bit more.

**Comment by jelbourn** [comment]
LGTM once you get the rebasing resolved

**Comment by hansl** [comment]
:+1: got the rebasing. Submitting once travis gives me the :green_heart:

**Pull request merged**